### PR TITLE
[16.0][MIG] pos_session_pay_invoice: Migration to version 16.0

### PR DESCRIPTION
**Pull request by carlos-lopez-tecnativa**
@tecnativa TT49807

### CODE REVIEW

File: pos_session_pay_invoice/readme/ROADMAP.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+* Cannot pay invoices in a different currency than that defined in the journal
+  associated to the payment method used to pay/collect payment.
+
+* Should depend on `pos_invoicing` but it requires a refactoring of `pos_invoicing`.
+  It will be improved when migrating to 13.0
```
Comment (legalsylvain) [inline pos_session_pay_invoice/readme/ROADMAP.rst]: This part is obsolete, as the invoices are now marked as paid.

**Review by legalsylvain — COMMENTED** [review]
Hi @carlos-lopez-tecnativa. 


Thanks for porting this module tecnically, I don't understand why 
- some button open a wizard that inherit from cash.pay.invoice,
- and in other button, it open a new model pos.box.cash.invoice.in / pos.box.cash.invoice.out.

could you elaborate ?

**Comment by carlos-lopez-tecnativa** [comment]
> Hi @carlos-lopez-tecnativa.
> 
> Thanks for porting this module tecnically, I don't understand why
> 
> * some button open a wizard that inherit from cash.pay.invoice,
> * and in other button, it open a new model pos.box.cash.invoice.in / pos.box.cash.invoice.out.
> 
> could you elaborate ?

Hi @legalsylvain,

In this module, I have only migrated and made minimal changes for now. However, if you consider it necessary, I can refactor it similarly to [this PR](https://github.com/OCA/account-payment/pull/743) to inherit from the new wizard `cash.pay.invoice` for customer/vendor invoices instead of using two separate wizards.

Before the referenced PR, these wizards (`cash.invoice.in`, `cash.invoice.out`) inherited from `cash.box.out`, but as it no longer exists, I created a single, more generic wizard called `cash.pay.invoice`. In this module, the wizard that depends on `cash.invoice.in` is updated to inherit from `cash.pay.invoice`. However, the wizards `pos.box.cash.invoice.out` and `pos.box.cash.invoice.in` do not inherit from any, so I didn't touch them.

Let me know if the refactor is welcome.

**Comment by legalsylvain** [comment]
Hi @carlos-lopez-tecnativa. 
Sorry. I didn't understood. I tested a little locally.
Functionaly, why there are two different ways to do the same thing ?

**Comment by carlos-lopez-tecnativa** [comment]
> Hi @carlos-lopez-tecnativa. Sorry. I didn't understood. I tested a little locally. Functionaly, why there are two different ways to do the same thing ?

@legalsylvain  That was the behavior in the previous version; there were several wizards to do each thing. Now, based on my previous comment and the referenced PR(https://github.com/OCA/account-payment/pull/743), we could simplify this. Let me know if that works for you. @pedrobaeza what do you think?

**Comment by pedrobaeza** [comment]
Yes, please simplify it.

**Comment by legalsylvain** [comment]
Agree with Pedro. Let me know when you want review.
Thanks for the work !

**Comment by carlos-lopez-tecnativa** [comment]
@pedrobaeza @legalsylvain I have updated the code to inherit functionality from the `cash.pay.invoice` wizard and removed the older wizards. Please review the changes and let me know if they work fine for you.

File: pos_session_pay_invoice/readme/ROADMAP.rst (hunk 1 above)
Comment (carlos-lopez-tecnativa) [reply pos_session_pay_invoice/readme/ROADMAP.rst]: removed

File: pos_session_pay_invoice/README.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+**Table of contents**
+
+.. contents::
+   :local:
+
+Configuration
+=============
+
+#.  Go to *Point of Sale / Configuration / Point of Sale* and activate the
```
Comment (pedrobaeza) [inline pos_session_pay_invoice/README.rst]: I think this text is outdated. There's no such option.

**Review by pedrobaeza — COMMENTED** [review]
I can't find the buttons specified in the README in the POS session.

File: pos_session_pay_invoice/README.rst (hunk 2 above)
Comment (carlos-lopez-tecnativa) [reply pos_session_pay_invoice/README.rst]: I updated the README. Please let me know if it is OK now.

**Comment by pedrobaeza** [comment]
I don't understand why limiting the invoice payment to cash methods. Someone can come to your physical store to pay an invoice, and choose to pay with credit card. I think that limitation should be gone.

**Comment by pedrobaeza** [comment]
And anyway, even adding a cash payment method, I don't find the buttons referred in the README

**Comment by carlos-lopez-tecnativa** [comment]
> And anyway, even adding a cash payment method, I don't find the buttons referred in the README

@pedrobaeza the buttons are here. Could you attach an image, please?
image

image

**Comment by pedrobaeza** [comment]
I was looking in the PoS session "frontend". In any moment the README mentions that you should go to the session backend form.

**Comment by carlos-lopez-tecnativa** [comment]
> I don't understand why limiting the invoice payment to cash methods. Someone can come to your physical store to pay an invoice, and choose to pay with credit card. I think that limitation should be gone.

This behavior is from the previous version. Do you want us to change it to allow the user to select any payment method available on POS? (But only if the payment method has a journal set.) Please confirm so I can adapt the code.

**Comment by pedrobaeza** [comment]
OK, I see the functionality very limited having to go to backend, so let's not invest more time in this. Please clarify the usage README though.

**Comment by carlos-lopez-tecnativa** [comment]
> OK, I see the functionality very limited having to go to backend, so let's not invest more time in this. Please clarify the usage README though.

Odoo, in the new version, has moved most functionalities to the frontend. This implies more effort to change it to the frontend. The task for now is to migrate the current module.
I have updated the readme. Please let me know if it is clearer now. Thanks.

**Review by pedrobaeza — APPROVED** [review]
Code and functional review

**Comment by carlos-lopez-tecnativa** [comment]
@legalsylvain, could you please review again?

**Comment by legalsylvain** [comment]
Hi @carlos-lopez-tecnativa. Thanks for the ping ! I just come back from hollidays today. I take a look this week. 

thanks !

**Comment by carlos-lopez-tecnativa** [comment]
> Hi @carlos-lopez-tecnativa. Thanks for the ping ! I just come back from hollidays today. I take a look this week.
> 
> thanks !

@legalsylvain I hope you had a great holiday! Would you kindly take another look when you have a chance? I'd appreciate any updates or feedback you may have.

**Review by ivantodorovich — APPROVED** [review]
code review LG

File: pos_session_pay_invoice/models/pos_order.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+from odoo import api, models
+from odoo.osv.expression import AND
+
+
+class PosOrder(models.Model):
+    _inherit = "pos.order"
+
+    @api.model
+    def search_paid_order_ids(self, config_id, domain, limit, offset):
```
Comment (legalsylvain) [inline pos_session_pay_invoice/models/pos_order.py]: Hi. Why this change in that module. I'm not sure to understand why this overload is interesting, but my question is why in this module ?

File: pos_session_pay_invoice/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#.  Go to *Point of Sale / Dashboard* and create and open or access to an
+    already open POS Session.
+#.  Open the POS Session form view on the Backend.
+#.  Press the button **Pay Invoice** to pay a Supplier Invoice or a Customer
+    Refund. It will be paid using Cash.
```
Comment (legalsylvain) [inline pos_session_pay_invoice/readme/USAGE.rst]:
```suggestion
#.  Press the button **Pay Supplier Invoice** to pay a Supplier Invoice or **Pay Refund** for a Customer
    Refund. It will be paid using Cash.
```

File: pos_session_pay_invoice/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+#.  Select **Collect Payment from Invoice** in to receive a payment of an
```
Comment (legalsylvain) [inline pos_session_pay_invoice/readme/USAGE.rst]:
```suggestion
#.  Select **Get Payment from Invoice** to receive a payment of an
```

**Review by legalsylvain — COMMENTED** [review]
Code review : 
`` 21 files changed, 219 insertions(+), 471 deletions(-)
``
- Nice simplification ! 
- a question regarding the pos_order.py file. 
- some minor adaptation in the readme files, to match with the UI.

Functional review : LGTM.

Thanks a lot for your work, and sorry for the delay.

File: pos_session_pay_invoice/models/pos_order.py (hunk 3 above)
Comment (carlos-lopez-tecnativa) [reply pos_session_pay_invoice/models/pos_order.py]: This code is present in V14. Please refer to this PR for the changes that were added: https://github.com/OCA/pos/pull/1006.
Let me know if this code is useful or if it should be removed.

Comment (carlos-lopez-tecnativa) [reply pos_session_pay_invoice/models/pos_order.py]: Reading the PR, the error is raised with orders created in this module, but now orders aren’t created; only `account.bank.statement.line` are created. So, I think the code is unused. Let me know what you think, please.

Comment (legalsylvain) [reply pos_session_pay_invoice/models/pos_order.py]: indeed, I think this part of code is now obsolete.

Comment (carlos-lopez-tecnativa) [reply pos_session_pay_invoice/models/pos_order.py]: Done, I removed the code.

Comment (legalsylvain) [reply pos_session_pay_invoice/models/pos_order.py]: thanks !

**Comment by legalsylvain** [comment]
Thanks !

**Comment by carlos-lopez-tecnativa** [comment]
I have no idea why the test failed. It works fine on my local environment. Does anyone have any ideas? The test that failed is in other modules.

image


image

**Comment by carlos-lopez-tecnativa** [comment]
I can see the error is general in the tour; other PRs have the same error in some OCA repos


https://github.com/OCA/calendar/actions/runs/10941425935/job/30376034658#step:8:159

https://github.com/OCA/pos/actions/runs/10921083459/job/30312371146#step:8:1086

https://github.com/OCA/e-commerce/actions/runs/10902558739/job/30254730888#step:8:306

**Comment by carlos-lopez-tecnativa** [comment]
@legalsylvain @pedrobaeza the PR is ready. The error in the tests is resolved in https://github.com/OCA/oca-ci/pull/78.
